### PR TITLE
feat: prototype manifest build

### PR DESCRIPTION
**Pull request by limeytexan**
## Proposed Changes

DO NOT MERGE

This patch implements a PoC of the manifest build which uses the instructions as found in the `[build]` section to build the package exactly as it would be built by manually invoking the same commands within a `flox activate` shell.

- sets $out in the environment
- invokes the build commands in a subshell (using bash)
- turns the $out directory into a Nix package with all outpath references replaced with the real $out
- wraps all `bin/*` commands with `$FLOX_ENV/activate`

Usage: flox-build [ pkgname ]

N.B. this patch utilises the (hitherto-unused) $FLOX_TURBO flag to invoke the provided command without an intervening shell, and in the process I tripped across and fixed a bug in the activate script.

## Release Notes

N/A

### CODE REVIEW

File: libexec/flox-build.mk [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  # Eventually derive version somehow, but hardcode it in the meantime.
+  $(eval _version = 0.0.0)
+  # Calculate name.
+  $(eval _name = $(_pname)-$(_version))
+
+  # Set temp outpath of same strlen as eventual package storePath using sha256sum
+  # derived from the package name, the current working directory and the $(FLOX_ENV)
+  # package to provide a stable random seed to avoid collisions.
+  $(eval _tmphash = $(shell ( \
+    echo $(_name) && pwd && realpath "$$FLOX_ENV") | sha256sum | head -c32))
```
Comment (tomberek) [inline libexec/flox-build.mk:47]: I've been adding FLOX_PROJECT as well. Imagine a Flox env being used along with different directories/projects.

File: libexec/flox-build.mk [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	FLOX_TURBO=1 out=$(_out) $(FLOX_ENV)/activate bash $$<
+	nix --extra-experimental-features nix-command \
+	  build -L --file __FLOX_CLI_OUTPATH__/libexec/build-manifest.nix \
+	    --argstr name "$(_name)" \
+	    --argstr flox-env "$(FLOX_ENV)" \
+	    --argstr install-prefix "$(_out)" \
+	    --out-link "result-$(_pname)" \
+	    --offline
+
+  # Type 2 "impure" build
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply libexec/flox-build.mk:89]: Thanks for highlighting - what option(s?) is missing? Is that `--no-sandbox`?

[the remark this answers is not in the document]
Comment (limeytexan) [reply libexec/flox-build.mk:89]: Ah .. I see the announcement in https://discourse.nixos.org/t/nix-2-8-0-released/18714 that explains what you mean. I've updated accordingly, but this also means we'll need to tweak the installer to set `extra-experimental-features = ca-derivations impure-derivations` in the `/etc/nix/flox.conf`, right?

[the remark this answers is not in the document]
Comment (limeytexan) [reply libexec/flox-build.mk:89]: Fixed with commit 76fb165a12686351e61e458ad3180b8a4f59aa40 - thanks again for the pointer!

Resolving conversation

File: libexec/build-manifest.nix [hunk 3]
```diff
@@ -25,21 +26,22 @@ in
...
       ''
       else ''
-               # If the build script is provided, then it's expected that we will
+        # If the build script is provided, then it's expected that we will
         # invoke it from within the sandbox to write directly to $out. The
         # choice of pure or impure mode occurs outside of this script as
         # the derivation is instantiated.
-               source $stdenv/setup
-               unpackPhase
+        export NIX_ENFORCE_PURITY=${if __impure then "0" else "1"}
```
Comment (tomberek) [inline libexec/build-manifest.nix:39]: Not sure about this conditional.

File: libexec/flox-build.mk [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+  # Identify result symlink basename.
+  $(eval _result = result-$(_pname))
+  # Eventually derive version somehow, but hardcode it in the meantime.
+  $(eval _version = 0.0.0)
+  # Calculate name.
+  $(eval _name = $(_pname)-$(_version))
+  # Short variable name for buildDependencies derived in the DEPENDS step.
+  $(eval $(_pvarname)_buildDeps_arg = $(strip \
+    $(if $($(_pvarname)_buildDeps),\
+      '["$(subst $(space),",$($(_pvarname)_buildDeps))"]')))
```
Comment (tomberek) [inline libexec/flox-build.mk:176]: something missing? a space?

**Comment by ghudgins** [comment]
closing as we are nearing PR on this

**Pull request closed without merging**